jumanji/environments/routing/pacman/utils.py

```diff
@@ -0,0 +1,470 @@
+        # are not allowed to backtrack.
+
+        is_in_tunnel = jnp.logical_or(
+            jnp.array_equal(valids, vert_col), jnp.array_equal(valids, hor_col)
+        )
+
+        def is_tunnel(
+            inputs: Tuple[chex.Array, chex.Array, int, chex.PRNGKey, int]
+        ) -> int:
+            """Repeat old action if in tunnel"""
+            _, _, ghost_action, _, _ = inputs
+            return ghost_action
+
+        def no_tunnel(
+            inputs: Tuple[chex.Array, chex.Array, int, chex.PRNGKey, int]
+        ) -> Any:
+            """Chose new action when at intersection"""
+            logits, actions, _, ghost_tunnel_key, _ = inputs
+            _, ghost_tunnel_key = jax.random.split(ghost_tunnel_key)
+            masked_logits = jnp.where(logits == 0, -1e9, logits)
+            # Softmax distribution
+            softmax_dist = nn.softmax(masked_logits)
+            weighted_actions = actions * logits
+            new_action = jax.random.choice(
+                ghost_tunnel_key, weighted_actions, p=softmax_dist
+            ).astype(int)
+            return new_action
+
+        inputs_no_tunnel = (
+            valid_actions,
+            act,
+            ghost_action,
+            ghost_tunnel_key,
+            ghost_num,
+        )
+
+        def start_over(
+            inputs: Tuple[chex.Array, chex.Array, int, chex.PRNGKey, int]
+        ) -> Any:
+            """If not in waiting mode then pick new action"""
+
+            chosen_action = jax.lax.cond(is_in_tunnel, is_tunnel, no_tunnel, inputs)
+            return jnp.squeeze(chosen_action)
+
+        def no_start(
+            inputs: Tuple[chex.Array, chex.Array, int, chex.PRNGKey, int]
+        ) -> int:
+            """If in waiting mode then use no-op"""
+            return 4
+
+        position = ghost_pos
+        chosen_action = jax.lax.cond(
+            ghost_start < 0, start_over, no_start, inputs_no_tunnel
+        )
+
+        # Use chosen action
+        move_left = lambda position: (position[1], position[0] - 1)
+        move_up = lambda position: (position[1] - 1, position[0])
+        move_right = lambda position: (position[1], position[0] + 1)
+        move_down = lambda position: (position[1] + 1, position[0])
+        no_op = lambda position: (position[1], position[0])
+
+        new_pos_row, new_pos_col = jax.lax.switch(
+            chosen_action,
+            [move_left, move_up, move_right, move_down, no_op],
+            position,
+        )
+
+        # If using the teleporter shift to the other side of the map
+        path = jnp.array([new_pos_col % state.y_size, new_pos_row % state.x_size])
+        path = jax.lax.cond(ghost_start <= 0, lambda: path, lambda: position)
+
+        return path, chosen_action
+
+    ghost_nums = jnp.array([0, 1, 2, 3])
+    ghost_init_targets = state.ghost_init_targets
+    ghost_paths, ghost_acts = jax.vmap(move, in_axes=(0, 0, 0, 0, None, 0, 0, 0, 0))(
```

Comment:
Just checking can the ghosts collide with eachother, what happens if they do? You're vmapping this move so it doesn't seem like you check for their collisions?

Reply:
The ghosts can't collide with each other, if they're on the same tile they'll just stack 👍